n3fit/src/n3fit/layers/observable.py

```diff
@@ -42,36 +43,104 @@ def __init__(self, fktable_data, fktable_arr, operation_name, nfl=14, **kwargs):
         super(MetaLayer, self).__init__(**kwargs)
 
         self.nfl = nfl
+        self.num_replicas = None  # set in build
+        self.compute_observable = None  # set in build
 
-        basis = []
+        all_bases = []
         xgrids = []
-        self.fktables = []
+        fktables = []
         for fkdata, fk in zip(fktable_data, fktable_arr):
             xgrids.append(fkdata.xgrid.reshape(1, -1))
-            basis.append(fkdata.luminosity_mapping)
-            self.fktables.append(op.numpy_to_tensor(fk))
+            all_bases.append(fkdata.luminosity_mapping)
+            fktables.append(op.numpy_to_tensor(fk))
+        self.fktables = fktables
 
         # check how many xgrids this dataset needs
         if is_unique(xgrids):
             self.splitting = None
         else:
             self.splitting = [i.shape[1] for i in xgrids]
 
-        # check how many basis this dataset needs
-        if is_unique(basis) and is_unique(xgrids):
-            self.all_masks = [self.gen_mask(basis[0])]
-            self.many_masks = False
-        else:
-            self.many_masks = True
-            self.all_masks = [self.gen_mask(i) for i in basis]
-
         self.operation = op.c_to_py_fun(operation_name)
         self.output_dim = self.fktables[0].shape[0]
 
+        if is_unique(all_bases) and is_unique(xgrids):
+            self.all_masks = [self.gen_mask(all_bases[0])]
+        else:
+            self.all_masks = [self.gen_mask(basis) for basis in all_bases]
+
+        self.masks = [self.compute_float_mask(bool_mask) for bool_mask in self.all_masks]
+
+    def build(self, input_shape):
+        self.num_replicas = input_shape[1]
+
+        # repeat the masks if necessary for fktables (if not, the extra copies
+        # will get lost in the zip)
+        masks = self.masks * len(self.fktables)
+        self.masked_fktables = [self.mask_fk(fk, mask) for fk, mask in zip(self.fktables, masks)]
+
+        super().build(input_shape)
+
     def compute_output_shape(self, input_shape):
         return (self.output_dim, None)
 
-    # Overridables
+    def compute_float_mask(self, bool_mask):
```

Comment:
why is this part of the layer? `self` is not used i think

Reply:
Agreed, moved it out.